### ISSUE TITLE
Complexities with serverside JS and SQL

### ISSUE THREAD

**Issue by PhantomYdn** [orientechnologies/orientdb#7565]
### OS: Ubuntu

Lets imagine that the following code is needed:

```javascript
var db = orient.getDatabase();
var list = ['writer', 'admin', 'reader'];
db.query("select from OUser where name in ?", list);
```

What's expected behavior? Will it return 3 users? No - it will return just one, because due to js boxing only 'writer' will be considered as first argument.

Lets modify slightly script to ensure that first argument is an array.

```javascript
var db = orient.getDatabase();
var list = ['writer', 'admin', 'reader'];
db.query("select from OUser where name in ?", [list]);
```

Result is Exception:

```java
java.lang.ClassCastException: key should be a String. It is java.lang.Integer instead.
     at jdk.nashorn.api.scripting.ScriptObjectMirror.checkKey(ScriptObjectMirror.java:885)
     at jdk.nashorn.api.scripting.ScriptObjectMirror.get(ScriptObjectMirror.java:382)
     at com.orientechnologies.orient.core.sql.parser.OPositionalParameter.bindFromInputParams(OPositionalParameter.java:46)
     at com.orientechnologies.orient.core.sql.parser.OPositionalParameter.toString(OPositionalParameter.java:30)
     at com.orientechnologies.orient.core.sql.parser.OBaseExpression.toString(OBaseExpression.java:55)
     at com.orientechnologies.orient.core.sql.parser.OInCondition.toString(OInCondition.java:96)
     at com.orientechnologies.orient.core.sql.parser.ONotBlock.toString(ONotBlock.java:58)
     at com.orientechnologies.orient.core.sql.parser.OAndBlock.toString(OAndBlock.java:61)
     at com.orientechnologies.orient.core.sql.parser.OOrBlock.toString(OOrBlock.java:73)
     at com.orientechnologies.orient.core.sql.parser.OWhereClause.toString(OWhereClause.java:47)
     at com.orientechnologies.orient.core.sql.parser.OSelectStatement.toString(OSelectStatement.java:170)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLAbstract.preParse(OCommandExecutorSQLAbstract.java:231)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:262)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:96)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:58)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:39)
     at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:3144)
     at com.orientechnologies.orient.core.sql.query.OSQLQuery.run(OSQLQuery.java:78)
     at com.orientechnologies.orient.core.sql.query.OSQLAsynchQuery.run(OSQLAsynchQuery.java:74)
     at com.orientechnologies.orient.core.sql.query.OSQLSynchQuery.run(OSQLSynchQuery.java:85)
     at com.orientechnologies.orient.core.query.OQueryAbstract.execute(OQueryAbstract.java:33)
     at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.query(ODatabaseDocumentTx.java:755)
     at com.orientechnologies.orient.core.command.script.OScriptDocumentDatabaseWrapper.query(OScriptDocumentDatabaseWrapper.java:88)
     at com.orientechnologies.orient.core.command.script.OScriptDocumentDatabaseWrapper.query(OScriptDocumentDatabaseWrapper.java:84)
```

OK. Lets try to use named parameters

```javascript
var db = orient.getDatabase();
var list = ['writer', 'admin', 'reader'];
db.query("select from OUser where name in :list", {"list" : list});
```

From user perspective (my perspective) it looks good, but - it returns nothing.

If you will try the following script it will return exception you can find down below.

```javascript
var db = orient.getDatabase();
var list = ['writer', 'admin', 'reader'];
db.query("select from OUser where name in [:list]", {"list" : list});
```

```java
com.orientechnologies.orient.core.exception.OSerializationException: Found invalid } character at position 13 of text "2": "reader"}. Ensure it is opened and closed correctly.
	DB name="Orienteer"
     at com.orientechnologies.orient.core.serialization.serializer.OStringSerializerHelper.parse(OStringSerializerHelper.java:372)
     at com.orientechnologies.orient.core.serialization.serializer.OStringSerializerHelper.parse(OStringSerializerHelper.java:305)
     at com.orientechnologies.orient.core.serialization.serializer.OStringSerializerHelper.smartSplit(OStringSerializerHelper.java:247)
     at com.orientechnologies.orient.core.sql.filter.OSQLFilterItemAbstract.<init>(OSQLFilterItemAbstract.java:58)
     at com.orientechnologies.orient.core.sql.filter.OSQLFilterItemField.<init>(OSQLFilterItemField.java:101)
     at com.orientechnologies.orient.core.sql.OSQLHelper.parseValue(OSQLHelper.java:241)
     at com.orientechnologies.orient.core.sql.OSQLHelper.parseValue(OSQLHelper.java:218)
     at com.orientechnologies.orient.core.sql.OSQLHelper.parseValue(OSQLHelper.java:214)
     at com.orientechnologies.orient.core.sql.filter.OSQLPredicate.convertCollectionItems(OSQLPredicate.java:357)
     at com.orientechnologies.orient.core.sql.filter.OSQLPredicate.extractConditionItem(OSQLPredicate.java:304)
     at com.orientechnologies.orient.core.sql.filter.OSQLPredicate.extractCondition(OSQLPredicate.java:201)
     at com.orientechnologies.orient.core.sql.filter.OSQLPredicate.extractConditions(OSQLPredicate.java:143)
     at com.orientechnologies.orient.core.sql.filter.OSQLPredicate.text(OSQLPredicate.java:95)
     at com.orientechnologies.orient.core.sql.filter.OSQLFilter.<init>(OSQLFilter.java:53)
     at com.orientechnologies.orient.core.sql.OSQLEngine.parseCondition(OSQLEngine.java:458)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:297)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:96)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:58)
     at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:39)
     at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:3144)
     at com.orientechnologies.orient.core.sql.query.OSQLQuery.run(OSQLQuery.java:78)
     at com.orientechnologies.orient.core.sql.query.OSQLAsynchQuery.run(OSQLAsynchQuery.java:74)
     at com.orientechnologies.orient.core.sql.query.OSQLSynchQuery.run(OSQLSynchQuery.java:85)
     at com.orientechnologies.orient.core.query.OQueryAbstract.execute(OQueryAbstract.java:33)
     at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.query(ODatabaseDocumentTx.java:755)
     at com.orientechnologies.orient.core.command.script.OScriptDocumentDatabaseWrapper.query(OScriptDocumentDatabaseWrapper.java:88)
     at com.orientechnologies.orient.core.command.script.OScriptDocumentDatabaseWrapper.query(OScriptDocumentDatabaseWrapper.java:84)
     at jdk.nashorn.internal.scripts.Script$38$\^eval\_.:program(<eval>:3)
     at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:637)
     at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
     at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
     at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:446)
     at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:403)
     at jdk.nashorn.api.scripting.NashornScriptEngine.evalImpl(NashornScriptEngine.java:399)
     at jdk.nashorn.api.scripting.NashornScriptEngine.eval(NashornScriptEngine.java:155)
```

At this moment I'm stuck. I can't understand how to pass properly array parameters without deep diving into hard coding java on javascript'. Please advise.

**Comment by luigidellaquila** [comment]
Hi @PhantomYdn 

I'd say it's a problem of conversion between js and java objects in the sql executor.
I'll take it a look and let you know

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @PhantomYdn 

A possible work-around is following:

```javascript
var db = orient.getDatabase();
var list = ['writer', 'admin', 'reader'];
var params = Java.to(list, "java.lang.Object[]");
return db.query("select from OUser where name in ?", params, null);
```

Unfortunately an automatic conversion cannot be implemented in the engine, as OrientDB v 2.2 is intended to work on both Java 7 and 8, with both Rhino and Nashorn, so I cannot hard code type conversions using Java 8 classes

I hope it helps

Thanks

Luigi
